### DISCUSSION TITLE
Maintainers, Collaborator and Contributors

### DISCUSSION THREAD

**Discussion by gbowne1** [gbowne1/codebooker#86] [General]
I would like to propose that this open source project have several maintainers, reviewers,  collaborators and contributors. My goal here is a minimum of one person per role, two with some coverage.

# Frontend

Write and maintain the frontend code.

Implement new features 

Debugging & testing browser and device specific bugs 

# Backend/Server

Maintain the code for the nodejs and expressjs server including the index.js / server.js

# Database

Maintain the mongodb code, schemas,  models, collection(s)

# Testing QA & QC

Writing and maintaining the e2e, unit and integration test code
  
Still need to decide which testing platforms we will use.  Jest and React  Testing Library are built in.  We can start there.

- Accessibility testing 

- Cross Browser Testing

- Device testing

# Documentation
 
Write, review and maintain application including international documentation.

# Deployment & DevOps
Make sure that builds run, containerized builds,  etc complete

Deployment to server

At this point in time I have been requiring a minimum of 5 successful merges before someone can be invited to be a collaborator and maintainer of the project.  

@LOGESH-B

**Comment by gbowne1** [comment]
I also think that at certain point we should require a minimum of 2 reviews on PRs from project maintainers in order to merge.  Certain types of frontend and backend, etc can be assigned via a CODEOWNER file in the project root folder/ directory.

**Reply by LOGESH-B** [reply]
Hii @gbowne1 
I am new to Open source contribution,I have done couple of projects Node,React,MongoDB and I  don't have any exposer in the areas of testing and devops..So you can assign accordingly🧑‍💻✌️

**Reply by gbowne1** [reply]
@LOGESH-B 

We can definitely assign testing and devops people as maintainers.

Theres still a few who have 3 and 4 succesfully merged PRs already.

The larger the project gets the easier it will be to maintain with at least 1 or 2 people per role and it would get documented. 

The contributions we have had so far have been pretty solid even if a few were pretty small.

**Comment by gbowne1** [comment]
@Vishwanath760970 @shivm29 @neeravbhaskarla @k-deepak04  @kabszac @Mayor-Rhymes @Yuva0 @Piyush-Deshmukh @kevinstewartsmith @javadkhangurbanov @goutham03062001 @atharvarekhawar @akansjain @777Prabhjot @giri-madhan @khalil-p @realrohitgurav @shyamtawli @victor20m

**Comment by k-deepak04** [comment]
@gbowne1 i can manage the fronend and backend task, but for now quite busy in open-source program, but will look the pr's issues etc once a week and help with them :)

**Reply by gbowne1** [reply]
Good to know.  I will be posting issues in the coming days.

**Reply by k-deepak04** [reply]
sure i'll be happy to contribute, collaborate and maintain the things :)

**Reply by gbowne1** [reply]
You already have contributed a little. 3 or 4 times I think. That's great. Thanks! :-)  Once you reach 5 successful merged PRs I will invite you to be a collaborator to the project. Feel free to contribute when you are able to.

**Reply by gbowne1** [reply]
We already have one person helping with the backend.  I think at least once a week is a solid target to work on something.

**Reply by k-deepak04** [reply]
okay i will, i guess i have contributed more than 5 but it's in your other repos also, i will contribute in this one too :)

**Reply by k-deepak04** [reply]
i will look into frontend and functionality aspect mainly.

**Reply by gbowne1** [reply]
Yes a few more than 5 total.  I chose 5 successful  merges in order to be invited to be collaborator as we have gotten a lot done and 5 gives us a chance to see how solid your contributions are.

The more people we have contributing and maintaining the app, the easier it will be.

**Comment by shyamtawli** [comment]
Hey I can look into frontend part.

**Reply by gbowne1** [reply]
Ok. @shyamtawli Plenty of Frontend stuff to do

**Comment by Vishwanath760970** [comment]
@gbowne1 I hope you are doing well, I see the backend implementation is started, good to see that, I've been busy lately with some interviews and office work, but I'll start looking into both frontend/backend tasks shortly.

**Reply by gbowne1** [reply]
Hi @Vishwanath760970 

All is well.

Yes, the backend has been started. Still a ways to go.  

Feel free to work on anything you like as you are you are able to. 

Thanks :-)

**Comment by gbowne1** [comment]
@BlackBond06 @alienqqe @atharvarekhawar @jchansen925 @victor20m @kabszac @javadkhangurbanov @giri-madhan @Jonnie-Dev @khalil-p @shyamtawli @Piyush-Deshmukh @ratishjain12 @neeravbhaskarla @realrohitgurav @shivm29 @shyamtawli @Yuva0

**Comment by BlackBond06** [comment]
@gbowne1, when I get to 5 successfully merged PRs, I would like to be part of maintaining the frontend part of the app.

**Reply by gbowne1** [reply]
Great.  This project has come a long ways.  We are almost to the point we will need QA/QC and to help @LOGESH-B and I review PRs to reduce the amount of errors.

Eventually I will require passing tests to be included with PRs.  Something that should be done as soon as possible.

**Comment by BlackBond06** [comment]
The project is really coming along very well I must say. I particularly love the enthusiasm shown by the contributors. I'm not really familiar with unit testing. Maybe I can make a research and see if I can help in that regard also. If you have resources on unit testing please share.

**Reply by gbowne1** [reply]
We will use Jest primarily.  It's the most common.  

There are documentation on the Jest site and hundreds of videos on YouTube.  Most of them are good.  

App.test.js is included 

Tests will be important for the longevity of the app.

Yes we have had a number of really good contributions.  I think this app could be very useful in a variety of ways.

This open source app is just as much everyone's app as it is mine.  Of course I could work on this alone but I felt community is where this app development is really at.  We can all make this app the best it can be.

I am also completely fine with people using this app to gain skills they would like to learn.

I am always open to ideas, thoughts, concerns, questions especially the further we get into development.

**Reply by gbowne1** [reply]
https://jestjs.io/docs/getting-started is the best place to start.  

App.test.js is included and this file comes out of the box when you run create react app and I think it's the most under used feature.  This file was meant to test App.js 

As an example Login testing would be most likely in Login.test.js and would be coded to test login functions.  

These files are normally kept in a /test or /tests or /__test, /__tests or even an  _ _ tests _ _ folder.  I prefer /tests but I would follow what the Jest docs say.

**Reply by gbowne1** [reply]
There is a couple videos I would suggest for learning Jest. I would start here, It is an hour long but provides a solid overview.

https://youtu.be/7r4xVDI2vho?si=lSwQKorVNRlqBhfT

**Reply by BlackBond06** [reply]
I definitely will check this out. Thank you

**Reply by gbowne1** [reply]
You're welcome. Glad to be of help.  😀

**Reply by LOGESH-B** [reply]
I am also started learning to writing test using jest✌️

**Comment by BlackBond06** [comment]
Thank you.

**Reply by gbowne1** [reply]
You're welcome :-)
